Userland/Libraries/LibGL/Shader.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2022, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <AK/StringBuilder.h>
+#include <LibGL/GLContext.h>
+
+namespace GL {
+
+GLuint GLContext::gl_create_shader(GLenum shader_type)
+{
+    // FIXME: Add support for GL_COMPUTE_SHADER, GL_TESS_CONTROL_SHADER, GL_TESS_EVALUATION_SHADER and GL_GEOMETRY_SHADER.
+    RETURN_VALUE_WITH_ERROR_IF(shader_type != GL_VERTEX_SHADER
+            && shader_type != GL_FRAGMENT_SHADER,
+        GL_INVALID_ENUM,
+        0);
+
+    GLuint shader_name;
+    m_shader_name_allocator.allocate(1, &shader_name);
+    auto shader = Shader::create(shader_type);
+    m_allocated_shaders.set(shader_name, shader);
+    return shader_name;
+}
+
+void GLContext::gl_delete_shader(GLuint shader)
+{
+    // "A value of 0 for shader will be silently ignored." (https://registry.khronos.org/OpenGL-Refpages/gl4/html/glDeleteShader.xhtml)
+    if (shader == 0)
+        return;
+
+    auto it = m_allocated_shaders.find(shader);
+    RETURN_WITH_ERROR_IF(it == m_allocated_shaders.end(), GL_INVALID_VALUE);
+
+    // FIXME: According to the spec, we should only flag the shader for deletion here and delete it once it is detached from all programs.
+    m_allocated_shaders.remove(it);
+    m_shader_name_allocator.free(shader);
+}
+
+void GLContext::gl_shader_source(GLuint shader, GLsizei count, GLchar const** string, GLint const* length)
+{
+    auto it = m_allocated_shaders.find(shader);
+    // FIXME: implement check "GL_INVALID_VALUE is generated if shader is not a value generated by OpenGL."
+    RETURN_WITH_ERROR_IF(it == m_allocated_shaders.end(), GL_INVALID_OPERATION);
+    RETURN_WITH_ERROR_IF(count < 0, GL_INVALID_VALUE);
+
+    it->value->clear_sources();
+    for (int i = 0; i < count; i++) {
+        if (!length || length[i] < 0)
```

Comment:
```suggestion
        if (length == nullptr || length[i] < 0)
```

Copying @Quaker762 's suggestion from the old PR.